OpenSim/Actuators/Thelen2003Muscle.cpp

```diff
@@ -762,110 +764,163 @@ Thelen2003Muscle::initMuscleState(const SimTK::State& s,
+            // unchanged
+            Fm = (ma*fal*fv + fpe)*fiso;
             FmAT = Fm*cosphi;
             Ft = fse*fiso;
-            ferr  = FmAT-Ft; 
-        
-            //Compute the partial derivative of the force error w.r.t. lce
-            dFm_dlce   = calcDFmDlce(lce, ma, fv, fiso, ofl);
-            dFmAT_dlce = calcDFmATDlce(lce,phi,cosphi,Fm, dFm_dlce, vol);
-            dFmAT_dlceAT= dFmAT_dlce*cosphi;
+            ferr = FmAT - Ft;
 
-            dFt_d_tl    = calcDFseDtl(tl, fiso, tsl);
-            dFt_d_lce   = calcDFseDlce(tl, lce, phi, cosphi,fiso, tsl, vol);
-            
+            if (h <= SimTK::SqrtEps ) {
+                break;
+            }
+            else
+                h = 0.5*h;;
+        }
 
-            dferr_d_lce = dFmAT_dlce - dFt_d_lce;
-
-            if(abs(ferr) > aSolTolerance){
-                //Take a full Newton Step                            
-                if(abs(dferr_d_lce) > SimTK::SignificantReal){
-                    delta_lce   = - ferr/dferr_d_lce;
-                    lce         = lce + delta_lce;                
-                }else{ //If we've stagnated, perturb the current solution            
-                    double perturbation = 
-                    2.0*((double)rand())/((double)RAND_MAX)-1.0;
-                    double lengthPerturbation = 
-                        0.5*perturbation*getOptimalFiberLength();
-                    lce = lce + lengthPerturbation;
-                }
+        ferrPrev = ferr;
+        lcePrev = lce;
+
+        //Update the partial derivative of the force error w.r.t. lce
+        dFm_dlce = calcDFmDlce(lce, ma, fv, fiso, ofl);
+        dFmAT_dlce = calcDFmATDlce(lce, phi, cosphi, Fm, dFm_dlce, vol);
+        dFmAT_dlceAT = dFmAT_dlce*cosphi;
+
+        dFt_d_tl = calcDFseDtl(tl, fiso, tsl);
+        dFt_d_lce = calcDFseDlce(tl, lce, phi, cosphi, fiso, tsl, vol);
+
+        //Update velocity level quantities: share the muscle velocity 
+        //between the tendon and the fiber according to their relative 
+        //stiffness:
+        //
+        //Fm = Ft at equilibrium
+        //Fm = Km*lceAT
+        //Ft = Kt*xt
+        //dFm_d_xm = Km*dlceAT + dKm_d_t*lceAT (assume dKm_d_t = 0)
+        //dFt_d_xt = Kt*dtl + dKt_d_t*dtl (assume dKt_d_t = 0)
+        //
+        //This is a heuristic. The above assumptions are necessary as 
+        //computing the partial derivatives of Km or Kt w.r.t. requires 
+        //acceleration level knowledge, which is not available in 
+        //general.
+
+        //Stiffness of the muscle is the stiffness of the tendon and the 
+        //fiber (along the tendon) in series
+
+        //the if statement here is to handle the special case when the
+        //negative stiffness of the fiber (which could happen in this
+        // model) is equal to the positive stiffness of the tendon.
+        if (abs(dFmAT_dlceAT + dFt_d_tl) > SimTK::SignificantReal
```

Comment:
The comment refers to the distribution of the muscle-tendon velocity to the fiber and tendon. Only the Millard2012EquilibriumMuscle had the a check and separate condition for evaluating the static fiber solution. The Thelen2003Muscle is able to solve for equilibrium without imposing the static assumption under all the conditions tested. I was not able to get the  Millard2012EquilibriumMuscle to do that. It does solve reliably with the static assumption imposed, which is it was doing before. I've only removed having to equilibrium methods and added the flag in the case we want to assume zero velocity.